### DISCUSSION TITLE
How to use planetscale from a docker setup

### DISCUSSION THREAD

**Discussion by Flyingmana** [planetscale/discussion#14] [Documentation] [answered]
As far as I understand the documentation, one needs to install the pscale CLI to establish a connection.

Iam a php developer.
I nowdays usually make use of docker or even host things in docker based environments.
Therefore it would be a great addition to also have an example how to establish this connection with docker.
And if its necessary to have it inside the app container, or if it can be a separate container which is used analog to any other database container.

**Comment by aquarapid** [comment]
@Flyingmana good point!  We'll work on an example of how to add the `pscale` CLI to a docker container, passing in auth, etc.

**Reply by Anders-PlanetScale** [reply]
Howdy,

The PlanetScale engineering team is working on building an official docker image but in the mean time I have written up a Dockerfile example which can be used to run a container that connects to a PlanetScale DB using `pscale connect` and exposes the connection port. 

[Dockerfile.txt](https://github.com/planetscale/beta/files/6549151/Dockerfile.txt)

To use it you save it in a local directory as Dockerfile and then run

`docker build .` from within the directory you saved it in.

You can then create an env_file (here is a template)
[env_file_example.txt](https://github.com/planetscale/beta/files/6549161/env_file_example.txt)

save your env_file in the same dir and edit the environment variables with your unique variable values.

Then you can run the docker image using the image ID your docker build command output OR you can find the ID using `docker image ls` and finding the recently created Image ID

`docker run -p 3306:3306 --env-file env_file <docker-image-ID>`

Feel free to build upon this example as you see fit and let us know if you have any questions.

I want to note that this is pretty close to what @Exellin shared previously, but is a bit more generic/slimmed down. Thanks for sharing your work Exellin!

**Reply by Anders-PlanetScale** [reply]
Oh I should note that my Dockerfile is assuming an i386 workstation you may need to change the pscale package based on your system.

**Comment by Exellin** [comment]
I have been comparing a bunch of setups recently and added a docker planetscale solution here: https://github.com/Exellin/node-graphql-comparisons/tree/master/entrypoints/docker-planetscale-mercurius-fastify-Graphql-js-Objection.  I am fairly new to docker and typescript in general, so probably lots of improvements to make.  My plan is to try out connecting [google cloud run](https://cloud.google.com/run) to planetscale.

The main things to take away were:

1. That I had to download and run tar on the executable from github with the following (Is there a way to always grab the latest?):

```
ADD https://github.com/planetscale/cli/releases/download/v0.36.0/pscale_0.36.0_linux_amd64.tar.gz .
RUN [ "tar", "zxvf", "pscale_0.36.0_linux_amd64.tar.gz", "-C", "/usr/bin" ]
```

2. That I had to create a .env file and pass in the following environment variables:
```
ARG PLANETSCALE_ORG
ENV PLANETSCALE_ORG=$PLANETSCALE_ORG

ARG PLANETSCALE_SERVICE_TOKEN_NAME
ENV PLANETSCALE_SERVICE_TOKEN_NAME=$PLANETSCALE_SERVICE_TOKEN_NAME

ARG PLANETSCALE_SERVICE_TOKEN
ENV PLANETSCALE_SERVICE_TOKEN=$PLANETSCALE_SERVICE_TOKEN
```

3. Had to run CMD with `node index.js` as a single command here:
```
CMD ["pscale", "connect", "node-graphql-comparisons", "main", "--execute", "node index.js"]
```

I run `docker build .`, then `docker run -p --env-file .env 4000:4000 ID`

**Reply by Anders-PlanetScale** [reply]
Howdy @Exellin,

I opened an issue to see if we can change our release package naming to be consistent across releases. If we can do that, it would allow you to pull the latest version of a package using something like;
` wget https://github.com/planetscale/cli/releases/latest/pscale_linux_386.deb`

**Comment by mscoutermarsh** [comment, marked as answer]
Hi all,

We now have a published docker image available. Instructions are in the CLI docs here: https://github.com/planetscale/cli#container-images.

If you have any feedback, please let us know here or open an issue in the CLI repo. 😄

**Reply by frankjrangel** [reply]
Hi! I have a question regarding the docker image, which path should I set for the volume so the container writes the .config to the host? right now it authenticates but it doesn't store the token anywhere

**Reply by DiademShoukralla** [reply]
Hello! Are you planning on publishing a docker image for arch arm64?  PlanetScale seems like a great solution for me, and works great in development, but my production environment is using Arm-based nodes.

**Reply by mscoutermarsh** [reply]
Hi Diadem,

How are you planning to use the CLI in production?

**Reply by DiademShoukralla** [reply]
Hi Mike! 
Thank you so much for the quick reply! From my understanding I need to use the pscale connect command to establish a proxy connection to the database. Please correct me if I am wrong.
I am using k8s to deploy my server's docker image. I am also planning on using Prisma to interact with the db!

**Reply by manigandham** [reply]
@DiademShoukralla
You don't need the CLI anymore. Planetscale supports standard mysql connections and you can create a connection string in the dashboard: https://docs.planetscale.com/concepts/connection-strings

**Reply by DiademShoukralla** [reply]
That is good to know! However I have tried doing that and keep receiving the following error whilst trying to connect 
```
server does not allow insecure connections, client must use SSL/TLS
```
I received this error while trying to use the connection string both in my local dev and in my production environment. Using the cli resolved the issue locally.

Any guidance would be greatly appreciated!

**Reply by manigandham** [reply]
@DiademShoukralla

The mysql connection needs to be secure (using TLS). Do you have `?sslaccept=strict` on the end of the connection string? 

![image](https://user-images.githubusercontent.com/457226/158913537-a4a227a1-5ebb-4f42-a6dd-a75a84955032.png)

**Reply by DiademShoukralla** [reply]
Thank you for the help, the problem is resolved now :)

**Reply by hhimanshu** [reply]
@mscoutermarsh , I tried running planetscale using docker and it fails for me. Do you know what I am missing? I ran the commands as per [`README`](https://github.com/planetscale/cli#readme)

```
Last login: Mon Jan 30 14:53:47 on ttys003
➜  ~ docker pull planetscale/pscale:latest

latest: Pulling from planetscale/pscale
8921db27df28: Pull complete 
c51aed88d9a2: Pull complete 
8b71c1a44b45: Pull complete 
Digest: sha256:0453efd81bc4ef4434e19926f2ee984036c909bc20c074a713c7fd8b9574b0a3
Status: Downloaded newer image for planetscale/pscale:latest
docker.io/planetscale/pscale:latest
➜  ~ docker run -e HOME=/tmp -v $HOME/.config/planetscale:/tmp/.config/planetscale --user $(id -u):$(id -g) --rm -it -p 3306:3306/tcp planetscale/pscale:latest
WARNING: The requested image's platform (linux/amd64) does not match the detected host platform (linux/arm64/v8) and no specific platform was requested
pscale is a CLI library for communicating with PlanetScale's API.

Usage:
  pscale [command]

Available Commands:
  audit-log      List audit logs
  auth           Login and logout via the PlanetScale API
  backup         Create, list, show, and delete branch backups
  branch         Create, delete, diff, and manage branches
  completion     Generate completion script for your shell
  connect        Create a secure connection to a database and branch for a local client
  data-imports   Create, list, and delete branch data imports
  database       Create, read, delete, and dump/restore databases
  deploy-request Create, review, diff, revert, and manage deploy requests
  help           Help about any command
  org            List, show, and switch organizations
  password       Create, list, and delete branch passwords
  region         List regions
  service-token  Create, list, and manage access for service tokens
  shell          Open a MySQL shell instance to a database and branch
  signup         Signup for a new PlanetScale account

Flags:
      --api-token string          The API token to use for authenticating against the PlanetScale API.
      --api-url string            The base URL for the PlanetScale API. (default "https://api.planetscale.com/")
      --config string             Config file (default is $HOME/.config/planetscale/pscale.yml)
      --debug                     Enable debug mode
  -f, --format string             Show output in a specific format. Possible values: [human, json, csv] (default "human")
  -h, --help                      help for pscale
      --no-color                  Disable color output
      --service-token string      Service Token for authenticating.
      --service-token-id string   The Service Token ID for authenticating.
      --version                   Show pscale version

Use "pscale [command] --help" for more information about a command.
➜  ~ 
```

**Reply by jgreet** [reply]
I'm pretty sure that's the CLI client's Docker image working properly. That's the same help text you'll see if you run `pscale` locally without any arguments. 

I think you reached out on Twitter as well. The longer answer is there's no local PlanetScale server you can use. It's probably simplest to use MySQL 8 instead which is what we do in house. You could run your own [Vitess cluster](https://vitess.io/) but that seems like a lot of overhead for a dev environment that still won't exactly match production.

**Reply by hhimanshu** [reply]
@jgreet , thanks a ton for your response. Just to make sure I understand everything, this is what I should do
1. Use `MySQL 8` docker image in local machine during development.
2. When ready to deploy, take the schema and data to PlanetScale using `pscale` (I don't know how yet)
Does this aligns with your advice?

**Reply by jgreet** [reply]
1. Yes, use MySQL 8.

2.  Most of the time you'll want to create a connection string and pass those details in to your application's database config as you would for any other database:
https://planetscale.com/docs/onboarding/connect-to-your-database#create-a-connection-string

You could use `pscale shell` (if you also have the MySQL CLI client installed) to quickly open a MySQL shell, or `pscale connect` to open a proxy on a local port for any other client you'd want to use. But you don't have to use pscale to connect to a database branch and I imagine most people don't at this point.

There was a point in the past where running `pscale connect` was the best way to connect to a branch on PlanetScale but that hasn't been true for a while now and is no longer recommended for production environments.

**Reply by hhimanshu** [reply]
This is huge help, thank you @jgreet

**Reply by fooforge** [reply]
@hhimanshu When you develop locally on your MySQL Community Server, keep in mind that Vitess has some [compatibility gaps](https://planetscale.com/docs/reference/mysql-compatibility) towards MySQL. We're continuously working on closing the gap and they're not a big deal if you develop a new app from scratch, but I wanted to drop this here for your awareness and to save your future self some refactoring time. ;-)

**Reply by hhimanshu** [reply]
@fooforge , thank you for the heads up! We will keep this in mind even though we are going to start an app from scratch
